# <font color=darkblue> Machine Learning model deployment with Flask framework</font>

## <font color=Blue>Used Cars Price Prediction Application</font>

### Objective:
1. To build a Machine learning regression model to predict the selling price of the used cars based on the different input features like fuel_type, kms_driven, type of transmission etc.
2. Deploy the machine learning model with the help of the flask framework.

### Dataset Information:
#### Dataset Source: https://www.kaggle.com/datasets/nehalbirla/vehicle-dataset-from-cardekho?select=CAR+DETAILS+FROM+CAR+DEKHO.csv
This dataset contains information about used cars listed on www.cardekho.com
- **Car_Name**: Name of the car
- **Year**: Year of Purchase
- **Selling Price (target)**: Selling price of the car in lakhs
- **Present Price**: Present price of the car in lakhs
- **Kms_Driven**: kilometers driven
- **Fuel_Type**: Petrol/diesel/CNG
- **Seller_Type**: Dealer or Indiviual
- **Transmission**: Manual or Automatic
- **Owner**: first, second or third owner


### 1. Import required libraries

In [1]:
## Import the libraries 

import numpy as np 
import pandas as pd 
import seaborn as sns 
import matplotlib.pyplot as plt 
%matplotlib inline 

from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
from sklearn.metrics import r2_score, accuracy_score, confusion_matrix

import warnings
warnings.filterwarnings('ignore')

import pickle

### 2. Load the dataset

In [2]:
# Loading the dataset

df = pd.read_csv('car data.csv')
df.head(5)

Car_Name  Year  Selling_Price  Present_Price  Kms_Driven Fuel_Type  \
0     ritz  2014           3.35           5.59       27000    Petrol   
1      sx4  2013           4.75           9.54       43000    Diesel   
2     ciaz  2017           7.25           9.85        6900    Petrol   
3  wagon r  2011           2.85           4.15        5200    Petrol   
4    swift  2014           4.60           6.87       42450    Diesel   

  Seller_Type Transmission  Owner  
0      Dealer       Manual      0  
1      Dealer       Manual      0  
2      Dealer       Manual      0  
3      Dealer       Manual      0  
4      Dealer       Manual      0

### 3. Check the shape and basic information of the dataset.

In [3]:
# Checking shape of the dataset

df.shape

(301, 9)

In [4]:
# Checking basic information about the dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Car_Name       301 non-null    object 
 1   Year           301 non-null    int64  
 2   Selling_Price  301 non-null    float64
 3   Present_Price  301 non-null    float64
 4   Kms_Driven     301 non-null    int64  
 5   Fuel_Type      301 non-null    object 
 6   Seller_Type    301 non-null    object 
 7   Transmission   301 non-null    object 
 8   Owner          301 non-null    int64  
dtypes: float64(2), int64(3), object(4)
memory usage: 21.3+ KB


- There are 301 rows & 9 fields in the dataset.
- No null values present in any of the fields.

### 4. Check for the presence of the duplicate records in the dataset? If present drop them

In [5]:
df[df.duplicated()]

Car_Name  Year  Selling_Price  Present_Price  Kms_Driven Fuel_Type  \
17    ertiga  2016           7.75          10.79       43000    Diesel   
93  fortuner  2015          23.00          30.61       40000    Diesel   

   Seller_Type Transmission  Owner  
17      Dealer       Manual      0  
93      Dealer    Automatic      0

- There are 2 duplicated rows in the dataset

In [6]:
# Removing the duplicated rows

df = df.drop_duplicates()

df.shape

(299, 9)

- After removal of 2 duplicate rows no of rows reduces from 301 to 299

### 5. Drop the columns which you think redundant for the analysis.

In [7]:
# Dropping the redundant columns

df = df.drop(columns=['Car_Name'], axis=1)
df.head()

Year  Selling_Price  Present_Price  Kms_Driven Fuel_Type Seller_Type  \
0  2014           3.35           5.59       27000    Petrol      Dealer   
1  2013           4.75           9.54       43000    Diesel      Dealer   
2  2017           7.25           9.85        6900    Petrol      Dealer   
3  2011           2.85           4.15        5200    Petrol      Dealer   
4  2014           4.60           6.87       42450    Diesel      Dealer   

  Transmission  Owner  
0       Manual      0  
1       Manual      0  
2       Manual      0  
3       Manual      0  
4       Manual      0

- We drop the 'Car_Name' column which is redundant as there can be many variation under the same Car Name

### 6. Extract a new feature called 'age_of_the_car' from the feature 'year' and drop the feature year

In [8]:
import datetime

# Geting age of the car
curr_year = datetime.date.today().year
df['age_of_the_car'] = curr_year - df['Year']

# Dropping 'Year' column after we get the age of the car
df = df.drop(columns=['Year'])

df.head()

Selling_Price  Present_Price  Kms_Driven Fuel_Type Seller_Type  \
0           3.35           5.59       27000    Petrol      Dealer   
1           4.75           9.54       43000    Diesel      Dealer   
2           7.25           9.85        6900    Petrol      Dealer   
3           2.85           4.15        5200    Petrol      Dealer   
4           4.60           6.87       42450    Diesel      Dealer   

  Transmission  Owner  age_of_the_car  
0       Manual      0              10  
1       Manual      0              11  
2       Manual      0               7  
3       Manual      0              13  
4       Manual      0              10

- A new column 'age_of_the_car' is extarcted from the Year column by the feature engineering methodology
- The 'Year' column is dropped after extracting the age as it is now redundant.

### 7. Encode the categorical columns

In [9]:
# Getting the ctegorical columns which needs encoding

categorical_columns = df.select_dtypes(include=['object']).columns.tolist()
categorical_columns


['Fuel_Type', 'Seller_Type', 'Transmission']

- We find presence of Three categorcal columns in the dataset - 'Fuel_Type', 'Seller_Type', 'Transmission'

In [10]:
# Getting the unique values in the 'Fuel_Type' column
df['Fuel_Type'].unique()


array(['Petrol', 'Diesel', 'CNG'], dtype=object)

In [11]:
# We employ manual encoding of the 'Fuel_Type' column

df['Fuel_Type'] = df['Fuel_Type'].replace({'Petrol':0,'Diesel':1,'CNG':2})
df.head()

Selling_Price  Present_Price  Kms_Driven  Fuel_Type Seller_Type  \
0           3.35           5.59       27000          0      Dealer   
1           4.75           9.54       43000          1      Dealer   
2           7.25           9.85        6900          0      Dealer   
3           2.85           4.15        5200          0      Dealer   
4           4.60           6.87       42450          1      Dealer   

  Transmission  Owner  age_of_the_car  
0       Manual      0              10  
1       Manual      0              11  
2       Manual      0               7  
3       Manual      0              13  
4       Manual      0              10

In [13]:
# Getting the unique values in the 'Seller_Type' column

df['Seller_Type'].unique()

array(['Dealer', 'Individual'], dtype=object)

In [14]:
# We employ manual encoding of the 'Seller_Type' column

df['Seller_Type'] = df['Seller_Type'].replace({'Dealer':0,'Individual':1})
df.head()

Selling_Price  Present_Price  Kms_Driven  Fuel_Type  Seller_Type  \
0           3.35           5.59       27000          0            0   
1           4.75           9.54       43000          1            0   
2           7.25           9.85        6900          0            0   
3           2.85           4.15        5200          0            0   
4           4.60           6.87       42450          1            0   

  Transmission  Owner  age_of_the_car  
0       Manual      0              10  
1       Manual      0              11  
2       Manual      0               7  
3       Manual      0              13  
4       Manual      0              10

In [15]:
# Getting the unique values in the 'Transmission' column

df['Transmission'].unique()

array(['Manual', 'Automatic'], dtype=object)

In [16]:
# We employ manual encoding of the 'Transmission' column

df['Transmission'] = df['Transmission'].replace({'Manual':0,'Automatic':1})
df.head()

Selling_Price  Present_Price  Kms_Driven  Fuel_Type  Seller_Type  \
0           3.35           5.59       27000          0            0   
1           4.75           9.54       43000          1            0   
2           7.25           9.85        6900          0            0   
3           2.85           4.15        5200          0            0   
4           4.60           6.87       42450          1            0   

   Transmission  Owner  age_of_the_car  
0             0      0              10  
1             0      0              11  
2             0      0               7  
3             0      0              13  
4             0      0              10

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 299 entries, 0 to 300
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Selling_Price   299 non-null    float64
 1   Present_Price   299 non-null    float64
 2   Kms_Driven      299 non-null    int64  
 3   Fuel_Type       299 non-null    int64  
 4   Seller_Type     299 non-null    int64  
 5   Transmission    299 non-null    int64  
 6   Owner           299 non-null    int64  
 7   age_of_the_car  299 non-null    int64  
dtypes: float64(2), int64(6)
memory usage: 21.0 KB


- After manual encoding of all the categorical columns we verify from the dataset information once again that all columns are now changed to numerical type ( float / int ).

### 8. Separate the target and independent features.

In [18]:
# Separating the independent columns & the target column into X & y

X = df.drop(columns=['Selling_Price'], axis=1)
y = df['Selling_Price']

In [19]:
X.head()

Present_Price  Kms_Driven  Fuel_Type  Seller_Type  Transmission  Owner  \
0           5.59       27000          0            0             0      0   
1           9.54       43000          1            0             0      0   
2           9.85        6900          0            0             0      0   
3           4.15        5200          0            0             0      0   
4           6.87       42450          1            0             0      0   

   age_of_the_car  
0              10  
1              11  
2               7  
3              13  
4              10

In [20]:
y.head()

0    3.35
1    4.75
2    7.25
3    2.85
4    4.60
Name: Selling_Price, dtype: float64

### 9. Split the data into train and test.

In [21]:
## Split the data in to train and test 

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3, random_state=0)
print(X_train.shape, X_test.shape)
print(y_train.shape, y_test.shape)

(209, 7) (90, 7)
(209,) (90,)


### 10. Build a Random forest Regressor model and check the r2-score for train and test.

In [23]:
# Loading the Random Forest Regressor model 
rf = RandomForestRegressor()

# Fitting the dataset in the model
rf.fit(X_train,y_train)

RandomForestRegressor()

In [24]:
## Accuracy check 

y_train_pred = rf.predict(X_train) 
y_test_pred = rf.predict(X_test)


accuracy_train = r2_score(y_train, y_train_pred)
accuracy_test = r2_score(y_test, y_test_pred)

print('Accuracy Train -> ', accuracy_train)
print('Accuracy Test -> ', accuracy_test)

Accuracy Train ->  0.9774141258217508
Accuracy Test ->  0.8937818327554976


### 11. Create a pickle file with an extension as .pkl

In [25]:
# Creating the pickle file

pickle.dump(rf, open('model.pkl','wb'))

### 12. Create new folder/new project in visual studio/pycharm that should contain the "model.pkl" file *make sure you are using a virutal environment and install required packages.*

### a) Create a basic HTML form for the frontend

Create a file **index.html** in the templates folder and copy the following code.

In [1]:
# index.html file 

<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>

    <div class="hero-image">
        <div class="hero-text">

            <h1>Used Car Price Predictor</h1>
            {% if prediction_text %}

            <h3>{{ prediction_text }}</h3>

            {% endif %}

        </div>
    </div>

    <style>
        body,
        html {
            height: 100%;
            margin: 0;
            font-family: Arial, Helvetica, sans-serif;
        }

        .hero-image {
            background-image: linear-gradient(rgba(0, 0, 0, 0.5), rgba(0, 0, 0, 0.5)), url('/static/image2.jpg');
            height: 40%;
            background-position: center;
            background-repeat: no-repeat;
            background-size: cover;
            position: relative;
        }

        .hero-text {
            text-align: center;
            position: absolute;
            top: 50%;
            left: 50%;
            transform: translate(-50%, -50%);
            color: white;


        }

        .hero-text h1 {
            font-size: 40px;
        }


        .hero-text h3 {
            color: yellow;
            padding: 16px;
            font-size: 24px;
            border: 1px solid white;
            border-radius: 3px;


        }
    </style>


    <div style="color:	rgb(0, 0, 0)">
        <form action="{{ url_for('predict')}}" method="post">
            <h2>Enter Car Details: </h2>

            <h3>Age of the car(In years)</h3>
            <input id="first" name="Age_of_the_car" type="number ">

            <h3>Present Showroom Price(In lakhs)</h3><br><input id="second" name="Present_Price" required="required">
            <h3>Kilometers Driven</h3><input id="third" name="Kms_Driven" required="required">
            <h3>Owner Type (0/1/2)</h3><br><input id="fourth" name="Owner" required="required">
            <h3>Fuel type</h3><br><select name="Fuel_Type" id="fuel" required="required">
                <option value="0">Petrol</option>
                <option value="1">Diesel</option>
                <option value="2">CNG</option>
            </select>
            <h3>Seller Type</h3><br><select name="Seller_Type" id="resea" required="required">
                <option value="0">Dealer</option>
                <option value="1">Individual</option>
            </select>
            <h3>Transmission type</h3><br><select name="Transmission" id="research" required="required">
                <option value="0">Manual Car</option>
                <option value="1">Automatic Car</option>
            </select>
            <br><br><button id="sub" type="submit ">Predict Selling Price</button>
            <br>


        </form>

    </div>

    <style>
        body {
            background-color: rgba(236, 172, 179, 0.411);
            text-align: center;
            padding: 0px;
            font-family: Helvetica;
        }


        #research {
            font-size: 18px;
            width: 200px;
            height: 23px;
            top: 23px;
        }

        #box {
            border-radius: 60px;
            border-color: 45px;
            border-style: solid;
            text-align: center;
            background-color: white;
            font-size: medium;
            position: absolute;
            width: 700px;
            bottom: 9%;
            height: 850px;
            right: 30%;
            padding: 0px;
            margin: 0px;
            font-size: 14px;
        }

        #fuel {
            width: 83px;
            height: 43px;
            text-align: center;
            border-radius: 14px;
            font-size: 20px;
        }

        #fuel:hover {
            background-color: white;
        }

        #research {
            width: 150px;
            height: 43px;
            text-align: center;
            border-radius: 14px;
            font-size: 18px;
        }

        #research:hover {
            background-color: white;
        }

        #resea {
            width: 99px;
            height: 43px;
            text-align: center;
            border-radius: 14px;
            font-size: 18px;
        }

        #resea:hover {
            background-color: white;
        }

        #sub {
            margin-block: 16px;
            background-color: Green;
            font-family: 'Helvetica' monospace;
            font-weight: bold;
            padding: 16px;
            text-align: center;
            border-radius: 20px;
            font-size: 18px;
            transition: all 0.1s ease-in;
            cursor: pointer;
            border: none;
            outline: none;
            background-color: transparent;
            color: green;
            border: 2px solid green;
        }

        #sub:hover {
            background-color: green;
            color: white;

        }

        #first {
            border-radius: 14px;
            height: 25px;
            font-size: 20px;
            text-align: center;
        }

        #second {
            border-radius: 14px;
            height: 25px;
            font-size: 20px;
            text-align: center;
        }

        #third {
            border-radius: 14px;
            height: 25px;
            font-size: 20px;
            text-align: center;
        }

        #fourth {
            border-radius: 14px;
            height: 25px;
            font-size: 20px;
            text-align: center;
        }
    </style>
</body>

</html>

### b) Create app.py file and write the predict function

In [ ]:
# app.py

from flask import Flask, request, jsonify, render_template
import numpy as np
import pickle

app = Flask(__name__)

model = pickle.load(open('model.pkl','rb'))

@app.route('/', methods=['GET','POST'])
def predict():
    
    if request.method == 'GET':
        return render_template('index.html')
    
    age_of_car = int(request.form['Age_of_the_car'])
    present_price = float(request.form['Present_Price'])
    kms_driven = float(request.form['Kms_Driven'])
    owner = int(request.form['Owner'])
    fuel_type = int(request.form['Fuel_Type'])
    seller_type = int(request.form['Seller_Type'])
    transmission = int(request.form['Transmission'])

    prediction = model.predict([[present_price,kms_driven,fuel_type, seller_type, transmission, owner, age_of_car ]])
    output = round(prediction[0],2)
    return render_template('index.html',prediction_text = f"Predicted Selling Price : {output} Lakhs" ) 
    

if __name__ == '__main__':
    app.run(debug=True)

### 13. Run the app.py python file which will render to index html page then enter the input values and get the prediction.

In [ ]:
# Run the command to run the app.py file and render the html page
python app.py

### Happy Learning :)